### ISSUE TITLE
OAuth error in Heroku custom domains

### ISSUE THREAD

**Issue by akhil-naidu** [forem/forem#13957]
I'm working on Forem in production and successfully got a Forem instance in Heroku. I configured s3, Cloudinary for image errors and finally attached my custom domain to it. Presently the instance is lying [here](https://app.leewardslope.com). 

As I have a working and persistent Forem, I started to configure more of it, like adding tags, configuring onboarding, and importantly different authentication methods. 

Everything went smooth, but when I was checking my custom authentication, I was receiving an error. So, I thought to go back to Heroku default URL and check the same, by changing all the URLs accordingly. This time I was successfully able to log in.

Issue: 
Call back URLs with custom domains in Heroku have some issue and I'm not sure what it is. I'm attaching the screenshot and Heroku log.

```
2021-06-10T09:55:40.469401+00:00 app[web.1]: E, [2021-06-10T09:55:40.469242 #34] ERROR -- : [20e08f23-6c20-4d46-9c69-9ef96f000c55] (github) Authentication failure! ActionController::InvalidAuthenticityToken: ActionController::InvalidAuthenticityToken, ActionController::InvalidAuthenticityToken
```
![Screenshot (52)](https://user-images.githubusercontent.com/11688537/121505246-4b798d00-ca00-11eb-978e-c466d3d4c230.png)

This is how I configured my OAuth in Github (Screenshot below) and I used the respective ID and KEY and placed them in the admin panel of my Forem.
![Screenshot (53)](https://user-images.githubusercontent.com/11688537/121506040-0013ae80-ca01-11eb-9c7f-ff4b7e125e5f.png)

**Comment by akhil-naidu** [comment]
@jdoss Do I need to provide any other information?

**Comment by Link2Twenty** [comment]
Have you added you OAuth keys to your `.env`?

https://docs.forem.com/backend/auth-github/

**Comment by akhil-naidu** [comment]
I used the admin gui to add the keys.

**Comment by akhil-naidu** [comment]
@Link2Twenty 

For making my point clear, 
1. I removed my custom domain and reverted it back to [Heroku domain](https://leewardslope-dev.herokuapp.com/)
2. Then I updated my APP_DOMAIN=leewardslope-dev.herokuapp.com
3. Lastly, in my Github developer settings of OAuth, I changed the Home URL and Callback URL according to this new URL.
4. The ENV values remain the same, I didn't interfere with it, as I knew I configured them properly.

Now, visit my [app](https://leewardslope-dev.herokuapp.com): https://leewardslope-dev.herokuapp.com, and try to signup with GITHUB, you can now do that without any issues.

While repeating the same procedure after changing to a custom domain, I'm facing issues. I will keep these changes intact until someone from @forem-team confirm them. So that I can proceed to change the settings into the custom domain and provide you with my Heroku logs.

**Comment by akhil-naidu** [comment]
After comparing both the Heroku domain and custom domain logs, I found a small but significant change in how Heroku Router works.

If I'm using Heroku Domain, by default, the protocol used was HTTPS, but this is not the same case while using a custom domain. Even though I'm using Cloudflare flexible universal SSL, the router within Heroku does not accept any SSL requests.

So, I thought to check this by manually adding an origin certificate(Cloudflares 15 year free origin SSL); but it claimed that I was not using a paid version at the end of this procedure.

Even though I'm testing this to write a guide to host Forem for free, I came out of my thing and upgraded my dyno to check whether this is the issue or not. Accordingly, I updated my dyno from free to the hobby and inserted my SSL certificate. 

Hola, everything started working as it intended to, and when I was looking at my logs, the Heroku router now has HTTPS protocol. `So, with this, I confirmed the issue was related to SSL and upgrading the dyno type would be the solution.`
---

Contrary to all this, I was away from my objective of creating a Forem for free of cost(with the working condition). I know that creating something like that will be of no use for anyone, but it will act as a helping hand for slow/non-tech developers in time. They can first test and configure everything within the free dyno, and once they feel comfortable configuring and getting the hang of the Admin panel, they can upgrade it to their needs.

So, I'm continuing to focus on resolving this issue and had some ideas to spare working with. So for anyone who is facing the same issue in free dyno, you know what to do(Upgrade your dyno to any paid version). For all those who want to know use all features in free dyno, I'm trying my best to resolve this issue within 5 days, even then if I were not able to finish, I would close this issue, and the only solution is the above-mentioned solution(to upgrade your dyno)
